File: typescript/src/image.ts
import { randomUUID } from "node:crypto";

export const ImageBuildOperationType = {
  ADD: "ADD",
  COPY: "COPY",
  ENV: "ENV",
  RUN: "RUN",
  WORKDIR: "WORKDIR",
} as const;

export type ImageBuildOperationType =
  (typeof ImageBuildOperationType)[keyof typeof ImageBuildOperationType];

export interface ImageBuildOperation {
  type: ImageBuildOperationType;
  args: string[];
  options: Record<string, string>;
}

export interface ImageOptions {
  name?: string;
  tag?: string;
  baseImage?: string | null;
}

function cloneOperation(op: ImageBuildOperation): ImageBuildOperation {
  return {
    type: op.type,
    args: [...op.args],
    options: { ...op.options },
  };
}

export class Image {
  readonly _id: string;
  readonly _name: string;
  readonly _tag: string;
  readonly _baseImage: string | null;
  readonly _buildOperations: ImageBuildOperation[];

  constructor();
  constructor(name: string, tag?: string, baseImage?: string | null);
  constructor(options: ImageOptions);
  constructor(
    nameOrOptions: string | ImageOptions = {},
    tag = "latest",
    baseImage: string | null = null,
  ) {
    this._id = randomUUID();
    this._buildOperations = [];

    if (typeof nameOrOptions === "string") {
      this._name = nameOrOptions;
      this._tag = tag;
      this._baseImage = baseImage;
      return;
    }

    this._name = nameOrOptions.name ?? "default";
    this._tag = nameOrOptions.tag ?? "latest";
    this._baseImage = nameOrOptions.baseImage ?? null;
  }

  get name(): string {
    return this._name;
  }

  get tag(): string {
    return this._tag;
  }

  get baseImage(): string | null {
    return this._baseImage;
  }

  get buildOperations(): ImageBuildOperation[] {
    return this._buildOperations.map(cloneOperation);
  }

  add(
    src: string,
    dest: string,
    options: Record<string, string> | undefined = undefined,
  ): this {
    return this._addOperation({
      type: ImageBuildOperationType.ADD,
      args: [src, dest],
      options: options == null ? {} : { ...options },
    });
  }

  copy(
    src: string,
    dest: string,
    options: Record<string, string> | undefined = undefined,
  ): this {
    return this._addOperation({
      type: ImageBuildOperationType.COPY,
      args: [src, dest],
      options: options == null ? {} : { ...options },
    });
  }

  env(key: string, value: string): this {
    return this._addOperation({
      type: ImageBuildOperationType.ENV,
      args: [key, value],
      options: {},
    });
  }

  run(
    commands: string | string[],
    options: Record<string, string> | undefined = undefined,
  ): this {
    return this._addOperation({
      type: ImageBuildOperationType.RUN,
      args: Array.isArray(commands) ? [...commands] : [commands],
      options: options == null ? {} : { ...options },
    });
  }

  workdir(directory: string): this {
    return this._addOperation({
      type: ImageBuildOperationType.WORKDIR,
      args: [directory],
      options: {},
    });
  }

  /**
   * Build this image as a sandbox template and register it.
   *
   * Materializes the image in a build sandbox, snapshots the filesystem,
   * and registers the snapshot as a named sandbox template.
   */
  async build(
    options: import("./sandbox-image.js").CreateSandboxImageOptions = {},
  ): Promise<Record<string, unknown>> {
    const { createSandboxImage } = await import("./sandbox-image.js");
    return createSandboxImage(this, options);
  }

  private _addOperation(op: ImageBuildOperation): this {
    this._buildOperations.push(op);
    return this;
  }
}

function renderOptions(options: Record<string, string>): string {
  const entries = Object.entries(options);
  if (entries.length === 0) {
    return "";
  }
  return ` ${entries.map(([key, value]) => `--${key}=${value}`).join(" ")}`;
}

function renderBuildOp(op: ImageBuildOperation): string {
  const options = renderOptions(op.options);
  if (op.type === ImageBuildOperationType.ENV) {
    return `ENV${options} ${op.args[0]}=${JSON.stringify(op.args[1])}`;
  }
  return `${op.type}${options} ${op.args.join(" ")}`;
}

export function dockerfileContent(image: Image): string {
  const lines = image.baseImage == null ? [] : [`FROM ${image.baseImage}`];
  lines.push(...image.buildOperations.map((op) => renderBuildOp(op)));
  return lines.join("\n");
}

File: src/tensorlake/image/image.py
from dataclasses import dataclass
from enum import Enum
from typing import Dict, List

from tensorlake.vendor.nanoid.nanoid import generate as nanoid_generate

_DEFAULT_BASE_IMAGE_NAME = "tensorlake/ubuntu-minimal"


class _ImageBuildOperationType(Enum):
    ADD = 1
    COPY = 2
    ENV = 3
    RUN = 4
    WORKDIR = 5


@dataclass
class _ImageBuildOperation:
    type: _ImageBuildOperationType
    args: List[str]
    options: Dict[str, str]


class Image:
    def __init__(
        self,
        name: str = "default",
        tag: str = "latest",
        base_image: str = _DEFAULT_BASE_IMAGE_NAME,
    ):
        # Used by ImageBuilder service to identify when different application
        # functions are using the same Image object.
        self._id: str = nanoid_generate()
        self._name: str = name
        self._tag: str = tag
        self._base_image: str = base_image
        self._build_operations: List[_ImageBuildOperation] = []

    @property
    def name(self) -> str:
        return self._name

    @property
    def tag(self) -> str:
        return self._tag

    def add(
        self, src: str, dest: str, options: Dict[str, str] | None = None
    ) -> "Image":
        return self._add_operation(
            _ImageBuildOperation(
                type=_ImageBuildOperationType.ADD,
                args=[src, dest],
                options={} if options is None else options,
            )
        )

    def copy(
        self, src: str, dest: str, options: Dict[str, str] | None = None
    ) -> "Image":
        return self._add_operation(
            _ImageBuildOperation(
                type=_ImageBuildOperationType.COPY,
                args=[src, dest],
                options={} if options is None else options,
            )
        )

    def env(self, key: str, value: str) -> "Image":
        return self._add_operation(
            _ImageBuildOperation(
                type=_ImageBuildOperationType.ENV, args=[key, value], options={}
            )
        )

    def run(
        self, commands: str | List[str], options: Dict[str, str] | None = None
    ) -> "Image":
        args = commands if isinstance(commands, list) else [commands]
        return self._add_operation(
            _ImageBuildOperation(
                type=_ImageBuildOperationType.RUN,
                args=args,
                options={} if options is None else options,
            )
        )

    def workdir(self, directory: str) -> "Image":
        return self._add_operation(
            _ImageBuildOperation(
                type=_ImageBuildOperationType.WORKDIR,
                args=[directory],
                options={},
            )
        )

    def build(
        self,
        *,
        registered_name: str | None = None,
        cpus: float = 2.0,
        memory_mb: int = 4096,
        disk_mb: int | None = None,
        builder_disk_mb: int | None = None,
        is_public: bool = False,
        docker_compat: bool = False,
        context_dir: str | None = None,
        verbose: bool = False,
    ) -> dict:
        """Build this image as a sandbox template and register it.

        Materializes the image in a build sandbox, snapshots the filesystem,
        and registers the snapshot as a named sandbox template.

        Args:
            registered_name: Name to register the image under. Defaults to ``self.name``.
            cpus: CPUs for the build sandbox (default 2.0).
            memory_mb: Memory for the build sandbox in MB (default 4096).
            disk_mb: Root disk size for the generated sandbox image in MB.
            builder_disk_mb: Root disk size for the temporary builder sandbox in MB.
            is_public: Make the registered image publicly accessible.
            docker_compat: Use Docker/BuildKit max compatibility mode (build
                is slower and uses more memory and disk space on builder
                sandbox).
            context_dir: Directory used to resolve relative COPY/ADD paths.
                When omitted, an empty build context is used (the generated
                Dockerfile needs no host files), so the current working
                directory is not uploaded. Pass this explicitly only when the
                image copies local files.
            verbose: If True, print build progress to stderr.

        Returns:
            The registered sandbox template response as a dict.
        """
        from .sandbox_builder import build_sandbox_image

        return build_sandbox_image(
            self,
            registered_name=registered_name,
            cpus=cpus,
            memory_mb=memory_mb,
            disk_mb=disk_mb,
            builder_disk_mb=builder_disk_mb,
            is_public=is_public,
            docker_compat=docker_compat,
            context_dir=context_dir,
            verbose=verbose,
        )

    def _add_operation(self, op: _ImageBuildOperation) -> "Image":
        self._build_operations.append(op)
        return self
